app/src/main/java/com/example/bottomnav/StructDecoder.java

```diff
@@ -6,42 +6,47 @@
 import java.util.Optional;
 
 public class StructDecoder implements DataDecoder {
-    private HashMap<String, DataDecoder> decodedPrimatives = new HashMap<>();
-    public ArrayList<VariableContents> variables;
-    private int size;
+    private ArrayList<Optional<DataDecoder>> decodedPrimatives;
```

Comment:
```suggestion
    private ArrayList<DataDecoder> decodedPrimitives;
```

I don't think this should be optional - there should not be a case where this is created with an unknown field. Partially because data sanitization should be handled at a higher level (before it becomes an internal variable) and partly because if there is an unknown field of unknown size that messes up any subsequent decoding.

Having your data sanitized before this also means you can clean up any downstream logic by removing the Optional tests.

Also, spelling.